clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+        NewRootValue->setProperty(kIsNull, Env.getBoolLiteralValue(false));
+
+        NewRootValue->setProperty(kIsNonnull, IsNonnull);
+    } else {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
+
+        NewRootValue->setProperty(kIsNull, IsNull);
+
+        Env.addToFlowCondition(A.makeNot(IsNonnull.formula()));
+        NewRootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    }
+}
+
+void matchNullptrExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    if (Env.getValue(*PrVar))
+      return;
+
+    // FIXME: Nullptr does not create a value
+    auto *RootValue = Env.createValue(PrVar->getType());
+    if (!RootValue) {
+       return;
+    }
+
+    RootValue->setProperty(kIsNull, Env.getBoolLiteralValue(true));
+    RootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+    Env.setValue(*PrVar, *RootValue);
+}
+
+void matchAddressofExpr(const Expr *expr, const MatchFinder::MatchResult &Result, Environment &Env) {
+    const auto *PrVar = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(PrVar != nullptr);
+
+    auto *RootValue = Env.createValue(PrVar->getType());
+    if (!RootValue) {
+       return;
+    }
+
+    RootValue->setProperty(kIsNull, Env.getBoolLiteralValue(false));
+    RootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(true));
+    Env.setValue(*PrVar, *RootValue);
+}
+
+void matchFunctionCallExpr(const Expr *fncall, const MatchFinder::MatchResult &Result, Environment &Env) {
+    // This is not necessarily a prvalue, since operators such as prefix ++ are lvalues.
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    // Initialize to (Unknown, Unknown)
+    if (Env.getValue(*Var))
+      return;
+
+    auto *RootValue = Env.createValue(Var->getType());
+    if (!RootValue)
+       return;
+    
+    initializeRootValue(*RootValue, Env);
+    setUnknownValue(*Var, *RootValue, Env);
+}
+
+ReverseNullDiagnoser::ResultType
+diagnoseCastExpr(const CastExpr *Stmt, const MatchFinder::MatchResult &Result,
+               const Environment &Env) {
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    Arena &A = Env.arena();
+
+    if (auto *RootValue = Env.getValue(*Var)) {
+        // FIXME: Hack
+        if (RootValue->getProperty(kIsNull) && RootValue->getProperty(kIsNonnull)) {
+            bool IsNull = !Env.flowConditionImplies(A.makeNot(getVal(kIsNull, *RootValue).formula()));
+            bool IsNonnull = !Env.flowConditionImplies(A.makeNot(getVal(kIsNonnull, *RootValue).formula()));
+
+            if (!IsNull && IsNonnull) {
+                return {{}, {Var->getBeginLoc()}};
+            }
+
+            if (IsNull && !IsNonnull) {
+                return {{Var->getBeginLoc()}, {}};
+            }
```

Comment:
What are the semantics of this return value? At the definition of the type alias, it is only said it's a pair of Loc vectors. How is it special if one part of the pair has elements, and how if the other. Is it possible that BOTH vectors will contain an element?